axlearn/common/checkpointer.py

```diff
@@ -619,16 +627,137 @@ class Config(Module.Config):
+    ):
         super().__init__(cfg, parent=parent)
+        if create_checkpointer:
+            if cfg.use_orbax:
+                self._checkpointer = OrbaxCheckpointer(cfg)
+            else:
+                self._checkpointer = StateStorageCheckpointer(cfg)
+
+    def __enter__(self):
+        self._checkpointer.enter()
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> Optional[bool]:
+        """Gracefully stops checkpointing, including waiting for async writes to finish."""
+        self._checkpointer.exit(exc_type, exc, traceback)
+
+    def save(
+        self, *, step: int, state: NestedTensor, evaler_summaries: Optional[Dict[str, Any]] = None
+    ):
+        """Saves checkpoint."""
+        self._checkpointer.save(step=step, state=state, evaler_summaries=evaler_summaries)
+
+    def restore(
+        self,
+        *,
+        step: Optional[int] = None,
+        state: Union[NestedTensor, NestedTensorSpec],
+    ):
+        """Restores checkpoint."""
+        return self._checkpointer.restore(step=step, state=state)
+
+    def stop(self):
+        """Gracefully stops checkpointing, including waiting for async writes to finish."""
+        self._checkpointer.exit()
+
+
+class OrbaxCheckpointer(Checkpointer):
+    """An implementation of Checkpointer using Orbax checkpoint."""
+
+    def __init__(self, cfg: Checkpointer.Config):
+        super().__init__(cfg, parent = None, create_checkpointer = False)
+        self._within_context = False
+        self._checkpoint_manager = CheckpointManager(
+            directory = cfg.dir,
+            options=CheckpointManagerOptions(
+                create=True,
+                save_interval_steps=cfg.save_policy.n,
+                max_to_keep=cfg.keep_last_n,
+                enable_async_checkpointing=True
+            )
+        )
+
+    def enter(self):
+        if self._within_context:
+            raise ValueError("Already in a context.")
+        self._within_context = True
+
+    def exit(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> Optional[bool]:
+        self.stop()
+        self._within_context = False
+
+    def save(
+        self, *, step: int, state: NestedTensor, evaler_summaries: Optional[Dict[str, Any]] = None
+    ):
+        self._checkpoint_manager.save(
+            step, args=ocp.args.StandardSave(item=state)
+        )
+
+    def restore(
+        self,
+        *,
+        step: Optional[int] = None,
+        state: Union[NestedTensor, NestedTensorSpec],
+    ) -> Tuple[Optional[int], NestedTensor]:
+        try:
+            transformed_state = jax.tree.map(self.transform_tensorspec, state)
+            restored_state = self._checkpoint_manager.restore(
+                step=self._checkpoint_manager.latest_step(),
```

Comment:
Redundant accesses of `latest_step`

Reply:
Also, you can specify step=None to automatically get the latest.

Reply:
I changed to `step=None` within the `restore`